src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/edit-polling-source.service.ts

```diff
@@ -0,0 +1,176 @@
+            case SetPollingSourceSection.READ: {
+                this.patchReadStep(sectionForm, editFormValue);
+                break;
+            }
+            case SetPollingSourceSection.MERGE: {
+                this.patchMergeStep(sectionForm, editFormValue);
+                break;
+            }
+        }
+    }
+
+    public getSetPollingSourceAsYaml(info: DatasetInfo) {
+        return this.appDatasetService
+            .requestDatasetHistory(info, this.historyLength, this.currentPage)
+            .pipe(
+                switchMap(
+                    () => this.appDatasetSubsService.onDatasetHistoryChanges,
+                ),
+                map((h: DatasetHistoryUpdate) => {
+                    this.history = h;
+                    const lastBlock = h.history.filter(
+                        (item: MetadataBlockFragment) =>
+                            item.event.__typename === "SetPollingSource",
+                    )[0];
+                    if (typeof lastBlock === "object") {
+                        return lastBlock.blockHash as string;
+                    }
+                    return h.history[0].blockHash as string;
+                }),
+                switchMap((hash: string) =>
+                    zip(
+                        this.blockService.onMetadataBlockAsYamlChanges,
+                        this.blockService.requestMetadataBlock(info, hash),
+                    ),
+                ),
+            );
+    }
+
+    private patchFetchStep(
+        sectionForm: FormGroup,
+        editFormValue: EditFormType,
+    ): void {
+        sectionForm.patchValue(editFormValue.fetch);
+        const eventTimeGroup = sectionForm.controls.eventTime as FormGroup;
+        const pattern = editFormValue.fetch.eventTime?.pattern;
+        const timestampFormat = editFormValue.fetch.eventTime?.timestampFormat;
+        eventTimeGroup.addControl("pattern", this.fb.control(pattern));
+        eventTimeGroup.addControl(
+            "timestampFormat",
+            this.fb.control(timestampFormat),
+        );
+        if (editFormValue.fetch.kind === FetchKind.URL) {
+            const headers = sectionForm.controls.headers as FormArray;
+            if ((headers.value as NameValue[]).length == 0)
+                editFormValue.fetch.headers?.forEach((item) => {
+                    headers.push(
+                        this.fb.group({
+                            name: [item.name],
+                            value: [item.value],
+                        }),
+                    );
+                });
+        } else if (editFormValue.fetch.kind === FetchKind.CONTAINER) {
+            const env = sectionForm.controls.env as FormArray;
+            editFormValue.fetch.env?.forEach((item) => {
+                env.push(
+                    this.fb.group({
+                        name: [item.name],
+                        value: [item.value],
+                    }),
+                );
+            });
```

Comment:
This code looks very similar to the code about "headers".
Maybe extract a helper method to initialize `NameValue[]`

Reply:
Ok, extract to helper method